lib/jerboa/client/worker.ex

```diff
@@ -96,161 +103,220 @@ defmodule Jerboa.Client.Worker do
+    :ok = UDP.send(socket, address, port, packet)
   end
 
-  @spec send_ind(state) :: :ok
-  defp send_ind(%{transaction: %{req: msg}} = state) do
-    socket = socket(state)
-    {address, port} = server(state)
-    :ok = UDP.send(socket, address, port, msg)
+  @spec handle_response(state, Params.t, Transaction.t) :: state
+  defp handle_response(state, params, transaction) do
+    handler = transaction.handler
+    {reply, creds, relay} = handler.(params, state.credentials, state.relay)
+    GenServer.reply(transaction.caller, reply)
+    %{state | credentials: creds, relay: relay}
   end
 
-  @spec request_allocation(state, retries_left :: pos_integer)
-    :: {result :: term, state}
-  defp request_allocation(state, retries_left) do
-    {result, new_state} =
-      state
-      |> Protocol.allocate_req()
-      |> send_req()
-      |> Protocol.eval_allocate_resp()
-    with :retry <- result,
-         n when n > 0 <- retries_left do
-      Logger.debug fn -> "Received error allocate response, retrying.." end
-      request_allocation(new_state, n - 1)
-    else
-      {:ok, _} ->
-        {result, update_lifetime_timer(new_state)}
-      _ ->
-        {result, new_state}
-    end
+  @spec setup_logger_metadata(state) :: any
+  defp setup_logger_metadata(%{socket: socket, server: server}) do
+    {:ok, port} = :inet.port(socket)
+    metadata = [jerboa_client: "#{inspect self()}:#{port}",
+                jerboa_server: Client.format_address(server)]
+    Logger.metadata(metadata)
   end
 
-  @spec update_lifetime_timer(state) :: state
-  defp update_lifetime_timer(state) do
-    timer_ref = state.lifetime_timer_ref
-    lifetime = state.lifetime
+  @spec update_allocation_timer(Relay.t) :: Relay.t
+  defp update_allocation_timer(relay) do
+    timer_ref = relay.timer_ref
+    lifetime = relay.lifetime
     if timer_ref do
       Process.cancel_timer timer_ref
     end
     if lifetime do
       new_ref = Process.send_after self(), :allocation_expired, lifetime * 1_000
-      %{state | lifetime_timer_ref: new_ref}
+      %{relay | timer_ref: new_ref}
     else
-      state
+      relay
     end
   end
 
-  @spec request_refresh(state, retries_left :: pos_integer)
-    :: {result :: term, state}
-  defp request_refresh(state, retries_left) do
-    {result, new_state} =
-      state
-      |> Protocol.refresh_req()
-      |> send_req()
-      |> Protocol.eval_refresh_resp()
-    with :retry <- result,
-         n when n > 0 <- retries_left do
-      Logger.debug fn -> "Received error refresh response, retrying.." end
-      request_refresh(new_state, n - 1)
-    else
-      :ok ->
-        {result, update_lifetime_timer(new_state)}
-      _ ->
-        {result, new_state}
+  ## Transactions
+
+  @spec add_transaction(state, Transaction.t) :: state
+  defp add_transaction(state, t) do
+    %{state | transactions: [t | state.transactions]}
+  end
+
+  @spec find_transaction(state, id :: binary) :: Transaction.t | nil
```

Comment:
If would be a bit simpler and potentially much faster to use map `transaction_id => Transaction.t`. What do you think about it?